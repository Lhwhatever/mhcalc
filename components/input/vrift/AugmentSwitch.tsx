import { Avatar, Checkbox, makeStyles, Typography } from '@material-ui/core'
import React from 'react'
import { textToKebabCase } from '../../../utils/str'
import { InputChangeEvent } from '../types'

export interface AugmentSwitchProps {
    label: string
    id?: string
    iconUri: string
    checked?: boolean
    onChange: (state: boolean) => void
}

const useStyles = makeStyles({
    augmentIcon: {
        width: 30,
        height: 30
    }
})

const AugmentSwitch = (props: AugmentSwitchProps): JSX.Element => {
    const { label, id, iconUri, checked, onChange } = props
    const labelId = id ?? 'augment-switch-' + textToKebabCase(label)

    const classes = useStyles()

    const handleCheckboxChange = (event: InputChangeEvent) => {
        onChange(event.target.checked)
    }

    return (
        <React.Fragment>
            <Checkbox inputProps={{ 'aria-labelledby': labelId }} checked={checked} onChange={handleCheckboxChange} />
            <Avatar src={iconUri} className={classes.augmentIcon} />
            <Typography variant="body1" id={labelId}>
                {label}
            </Typography>
        </React.Fragment>
    )
}

export default AugmentSwitch
