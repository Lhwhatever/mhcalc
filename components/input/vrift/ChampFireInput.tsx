import React from 'react'
import { Switch, FormControlLabel } from '@material-ui/core'
import { InputChangeEvent } from '../types'

export interface ChampFireInputProps {
    value?: boolean
    onChange?: (state: boolean) => void
}

const ChampFireInput = (props: ChampFireInputProps): JSX.Element => {
    const { value, onChange = () => undefined } = props

    const handleChange = (event: InputChangeEvent) => {
        onChange(event.target.checked)
    }

    return (
        <FormControlLabel
            control={<Switch checked={value} onChange={handleChange} inputProps={{ role: 'switch checkbox' }} />}
            label="Champion's Fire"
        />
    )
}

export default ChampFireInput
