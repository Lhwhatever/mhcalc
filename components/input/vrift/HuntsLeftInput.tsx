import { TextField } from '@material-ui/core'
import React from 'react'
import { InputEvent } from '../types'

export interface HuntsLeftInputProps {
    fullWidth?: boolean
    value?: number
    onChange?: (event: InputEvent) => void
}

const HuntsLeftInput = (props: HuntsLeftInputProps): JSX.Element => {
    return <TextField label="Hunts Left" type="number" variant="outlined" {...props} />
}

export default HuntsLeftInput
