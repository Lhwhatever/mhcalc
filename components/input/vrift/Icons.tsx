import { Avatar, withStyles } from '@material-ui/core'
import React from 'react'

export const SyncIcon = (): JSX.Element => (
    <Avatar
        alt="Sync"
        src="https://www.mousehuntgame.com/images/ui/hud/rift_valour/power_up_stamina.png"
        variant="rounded"
    />
)

export const SpeedIcon = (): JSX.Element => (
    <Avatar
        alt="Speed"
        src="https://www.mousehuntgame.com/images/ui/hud/rift_valour/power_up_long_stride.png"
        variant="rounded"
    />
)

export const SiphonIcon = (): JSX.Element => (
    <Avatar
        alt="Siphon"
        src="https://www.mousehuntgame.com/images/ui/hud/rift_valour/power_up_siphon.png"
        variant="rounded"
    />
)

export interface ChampFireIconProps {
    state?: boolean
}

const champFireStyles = {
    default: {
        width: '100%',
        paddingBottom: '100%',
        backgroundImage: 'url(https://www.mousehuntgame.com/images/ui/hud/rift_valour/fuel.png)',
        backgroundPosition: '0 0',
        backgroundSize: '100%',
        borderRadius: '50%'
    },
    active: {
        backgroundPosition: '0% 100%'
    }
}

export const ChampFireIcon = withStyles(champFireStyles)(
    ({ state = false }: ChampFireIconProps): JSX.Element => (
        <div
            style={{
                ...champFireStyles.default,
                ...(state ? champFireStyles.active : undefined)
            }}
            role="img"
            aria-label="champion's fire"
        />
    )
)
