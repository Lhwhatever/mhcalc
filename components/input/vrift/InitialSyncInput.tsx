import { MenuItem, TextField } from '@material-ui/core'
import React from 'react'
import { InputChangeEvent } from '../types'

export type InitialSync = 40 | 50 | 60 | 70 | 80 | 90 | 100

export interface SyncInputProps {
    fullWidth?: boolean
    value?: InitialSync
    onChange?: (event: InputChangeEvent) => void
}

const InitialSyncInput = (props: SyncInputProps): JSX.Element => {
    return (
        <TextField select label="Sync" variant="outlined" defaultValue={100} {...props}>
            {Array(7)
                .fill(0)
                .map((_, i) => {
                    const sync = i * 10 + 40
                    return (
                        <MenuItem key={i} value={sync}>
                            Lvl. {i + 1} ({sync} Sync)
                        </MenuItem>
                    )
                })}
        </TextField>
    )
}

export default InitialSyncInput
