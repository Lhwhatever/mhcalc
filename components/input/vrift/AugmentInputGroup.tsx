import React from 'react'
import { Box, Typography, makeStyles } from '@material-ui/core'
import AugmentSwitch from './AugmentSwitch'
import { useSelector, useDispatch } from 'react-redux'
import { RootState } from '../../../redux/rootReducer'
import { Augments, updateAugment } from '../../../redux/ducks/vrift/simInput'
import { camelCaseToTitleCase } from '../../../utils/str'

const useStyles = makeStyles((theme) => ({
    switchboxRoot: {
        display: 'grid',
        gridTemplateColumns: 'auto auto 1fr',
        gridColumnGap: theme.spacing(1),
        alignItems: 'center'
    }
}))

interface AugmentEntry {
    key: keyof Augments
    icon: string
}

const augmentSwitches: AugmentEntry[] = [
    { key: 'sigilHunter', icon: 'highway_robbery' },
    { key: 'secretResearch', icon: 'champion_pillage' },
    { key: 'superSiphon', icon: 'secret_passage' },
    { key: 'ultimateUmbra', icon: 'tower_ultimatum' },
    { key: 'stringStepping', icon: 'string_stepping' }
]

const AugmentInputGroup = (): JSX.Element => {
    const classes = useStyles()

    const enabledAugments = useSelector((state: RootState) => state.vrift.simInput.augments)
    const dispatch = useDispatch()

    const createAugmentStateHandler = (key: keyof Augments) => (state: boolean) => {
        dispatch(updateAugment({ target: key, state }))
    }

    return (
        <Box>
            <Typography variant="h4">Augments</Typography>
            <div className={classes.switchboxRoot}>
                {augmentSwitches.map((augment: AugmentEntry) => (
                    <AugmentSwitch
                        key={augment.key}
                        label={camelCaseToTitleCase(augment.key)}
                        iconUri={`https://www.mousehuntgame.com/images/ui/hud/rift_valour/augments/${augment.icon}.png`}
                        checked={enabledAugments[augment.key] ?? false}
                        onChange={createAugmentStateHandler(augment.key)}
                    />
                ))}
            </div>
        </Box>
    )
}

export default React.memo(AugmentInputGroup)
