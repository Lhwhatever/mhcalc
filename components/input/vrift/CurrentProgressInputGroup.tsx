import { Box, Button, makeStyles, Tooltip, Typography } from '@material-ui/core'
import InfoIconOutlined from '@material-ui/icons/InfoOutlined'
import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { updateHuntsLeft, updateInitialSync, updateSteps } from '../../../redux/ducks/vrift/simInput'
import { RootState } from '../../../redux/rootReducer'
import { InputChangeEvent } from '../types'
import HuntsLeftInput from './HuntsLeftInput'
import { SyncIcon } from './Icons'
import InitialSyncInput, { InitialSync } from './InitialSyncInput'
import StepsInput from './StepsInput'

const useStyles = makeStyles((theme) => ({
    inputGrid: {
        display: 'grid',
        gridTemplateColumns: '1fr 3fr 3rem',
        gridGap: theme.spacing(1)
    },
    inputIconLabel: {
        gridColumn: '1 / span 1',
        display: 'flex',
        flexDirection: 'row-reverse',
        alignItems: 'center'
    },
    inputField: {
        gridColumn: '2 / span 1',
        '& .MuiTextField-root': {
            minWidth: 166
        }
    },
    inputAdornment: {
        gridColumn: '3 / span 1',
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center'
    }
}))

const CurrentProgressInputGroup = (): JSX.Element => {
    const classes = useStyles()

    const { initialSync, huntsLeft, steps } = useSelector((state: RootState) => state.vrift.simInput)
    const dispatch = useDispatch()

    const handleInitialSyncChange = (event: InputChangeEvent) => {
        dispatch(updateInitialSync(parseInt(event.target.value) as InitialSync))
    }

    const handleHuntsLeftChange = (huntsLeft: number | undefined) => {
        dispatch(updateHuntsLeft(huntsLeft))
    }

    const handleStepsChange = (steps: number | undefined) => {
        dispatch(updateSteps(steps))
    }

    const handleResetToStart = () => {
        handleHuntsLeftChange(initialSync)
        handleStepsChange(0)
    }

    return (
        <Box>
            <Typography variant="h4">Current Progress</Typography>
            <Box className={classes.inputGrid} mt={2}>
                <div className={classes.inputIconLabel}>
                    <SyncIcon />
                </div>
                <div className={classes.inputField}>
                    <InitialSyncInput
                        fullWidth
                        value={initialSync}
                        onChange={handleInitialSyncChange}
                        id="initial-sync-input"
                    />
                </div>
                <div className={classes.inputAdornment}>
                    <Tooltip title="If you are upgrading Sync during the run, use the final Sync level.">
                        <InfoIconOutlined />
                    </Tooltip>
                </div>
                <div className={classes.inputField}>
                    <HuntsLeftInput
                        fullWidth
                        value={huntsLeft}
                        onChange={handleHuntsLeftChange}
                        id="hunts-left-input"
                    />
                </div>
                <div className={classes.inputField}>
                    <StepsInput fullWidth value={steps} onChange={handleStepsChange} id="steps-input" />
                </div>
            </Box>
            <Box mt={2} display="flex" flexDirection="row-reverse">
                <Button id="btn-reset-run" variant="outlined" onClick={handleResetToStart}>
                    Reset Progress
                </Button>
            </Box>
        </Box>
    )
}

export default CurrentProgressInputGroup
