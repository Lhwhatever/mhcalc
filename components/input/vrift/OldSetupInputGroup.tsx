import { Box, BoxProps, makeStyles, MenuItem, TextField, Typography } from '@material-ui/core'
import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { updateChampFire, updateLuck, updatePower, updateSiphon, updateSpeed } from '../../../redux/ducks/vrift/setups'
import { Siphon, Speed } from '../../../redux/ducks/vrift/stats'
import { RootState } from '../../../redux/rootReducer'
import { AttBonusIcon, LuckIcon, PowerIcon } from '../../icons/TrapStatIcons'
import { AttBonusInput, LuckInput, PowerInput } from '../TrapStatsInput'
import { InputChangeEvent } from '../types'
import ChampFireInput from './ChampFireInput'
import { ChampFireIcon, SiphonIcon, SpeedIcon } from './Icons'
import SiphonInput from './SiphonInput'
import SpeedInput from './SpeedInput'

const useStyles = makeStyles((theme) => ({
    mobileSetupInputRoot: {
        display: 'grid',
        margin: theme.spacing(1),
        gridGap: theme.spacing(1),
        gridTemplateColumns: 'auto 1fr',
        alignItems: 'center'
    }
}))

interface SetupTab {
    value: string
    label: string
}

const tabs: SetupTab[] = [
    { value: 'e1', label: 'Eclipse' },
    { value: 'r1', label: 'Floors 1-7' },
    { value: 'r9', label: 'Floors 9-15' },
    { value: 'r17', label: 'Floors 17-23' },
    { value: 'r25', label: 'Floors 25+' }
]

const ActiveComponentCompact = (): JSX.Element => {
    const classes = useStyles()

    const [activeTab, setActiveTab] = React.useState(tabs[0].value)

    const handleActiveTabChange = (event: InputChangeEvent) => {
        setActiveTab(event.target.value)
    }

    const setupType = activeTab.charAt(0) === 'e' ? 'eclipse' : 'regular'
    const setupLevel = parseInt(activeTab.slice(1))

    const { power, luck, speed, champFire } = useSelector(
        (state: RootState) => state.vrift.setups[setupType][setupLevel]
    )
    const siphon = useSelector((state: RootState) =>
        setupType === 'eclipse' ? state.vrift.setups.eclipse[setupLevel].siphon : undefined
    )

    const dispatch = useDispatch()

    const handlePowerChange = (power: number | undefined) => {
        dispatch(updatePower(setupType, power, setupLevel))
    }

    const handleLuckChange = (luck: number | undefined) => {
        dispatch(updateLuck(setupType, luck, setupLevel))
    }

    const handleSpeedChange = (speed: Speed) => {
        dispatch(updateSpeed(setupType, speed, setupLevel))
    }

    const handleSiphonChange = (siphon: Siphon) => {
        dispatch(updateSiphon(siphon, setupLevel))
    }

    const handleChampFireChange = (champFire: boolean) => {
        dispatch(updateChampFire(setupType, champFire, setupLevel))
    }

    return (
        <React.Fragment>
            <Box m={1}>
                <TextField
                    select
                    fullWidth
                    variant="outlined"
                    label="Setup"
                    id="setup-select"
                    value={activeTab}
                    onChange={handleActiveTabChange}
                >
                    {tabs.map(({ value, label }) => (
                        <MenuItem key={value} value={value}>
                            {label}
                        </MenuItem>
                    ))}
                </TextField>
            </Box>
            <div className={classes.mobileSetupInputRoot}>
                <div>
                    <PowerIcon />
                </div>
                <div>
                    <PowerInput fullWidth id="power-input" value={power ?? undefined} onChange={handlePowerChange} />
                </div>
                <div>
                    <LuckIcon />
                </div>
                <div>
                    <LuckInput fullWidth id="luck-input" value={luck ?? undefined} onChange={handleLuckChange} />
                </div>
                <div>
                    <AttBonusIcon />
                </div>
                <div>
                    <AttBonusInput disabled fullWidth id="att-bonus-input" />
                </div>
                <div>
                    <SpeedIcon />
                </div>
                <div>
                    <SpeedInput fullWidth id="speed-input" value={speed} onChange={handleSpeedChange} />
                </div>
                {setupType === 'eclipse' && (
                    <React.Fragment>
                        <div>
                            <SiphonIcon />
                        </div>
                        <div>
                            <SiphonInput fullWidth id="siphon-input" value={siphon!} onChange={handleSiphonChange} />
                        </div>
                    </React.Fragment>
                )}
                <div>
                    <ChampFireIcon state={champFire} />
                </div>
                <div>
                    <ChampFireInput value={champFire} onChange={handleChampFireChange} />
                </div>
            </div>
        </React.Fragment>
    )
}

const SetupInputGroup = (props: BoxProps): JSX.Element => {
    return (
        <Box {...props}>
            <Typography variant="h4">Setup</Typography>
            <ActiveComponentCompact />
        </Box>
    )
}

export default SetupInputGroup
