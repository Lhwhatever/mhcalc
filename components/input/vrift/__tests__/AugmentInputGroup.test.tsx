import React from 'react'
import { Augments, updateAugment } from '../../../../redux/ducks/vrift/simInput'
import mockStore from '../../../../utils/testing/mockStore'
import { createRenderWithRedux, fireEvent, screen } from '../../../../utils/testing/test'
import AugmentInputGroup from '../AugmentInputGroup'

describe('AugmentInputGroup test', () => {
    const initialAugmentState: Augments = { superSiphon: true }
    const store = mockStore({
        vrift: { simInput: { augments: initialAugmentState } }
    })

    const dispatchSpy = jest.spyOn(store, 'dispatch')
    const augmentNames = [/sigil hunter/i, /secret research/i, /super siphon/i, /ultimate umbra/i, /string stepping/i]

    beforeEach(() => {
        createRenderWithRedux(store)(<AugmentInputGroup />)
    })

    it('should have a header', () => {
        expect(screen.getByRole('heading', { name: /augments/i })).toBeInTheDocument()
    })

    it('should have seven checkbox inputs', () => {
        augmentNames.forEach((augment) => {
            expect(screen.getByLabelText(augment)).toBeInTheDocument()
        })
    })

    it('should show the state of the checkboxes', () => {
        ;[0, 1, 3, 4].forEach((i) => {
            expect(screen.getByLabelText(augmentNames[i])).not.toBeChecked()
        })

        expect(screen.getByLabelText(augmentNames[2])).toBeChecked()
    })

    it('should dispatch changes to the redux store', () => {
        fireEvent.click(screen.getByLabelText(augmentNames[0]))
        expect(dispatchSpy).lastCalledWith(updateAugment({ target: 'sigilHunter', state: true }))

        fireEvent.click(screen.getByLabelText(augmentNames[2]))
        expect(dispatchSpy).lastCalledWith(updateAugment({ target: 'superSiphon', state: false }))
    })

    it('should match snapshot', () => {
        expect(screen).toMatchSnapshot()
    })
})
