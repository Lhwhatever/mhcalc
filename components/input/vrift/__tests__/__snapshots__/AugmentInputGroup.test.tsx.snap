// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AugmentInputGroup test should match snapshot 1`] = `
<Styled(MuiBox)>
  <WithStyles(ForwardRef(Typography))
    variant="h4"
  >
    Augments
  </WithStyles(ForwardRef(Typography))>
  <div
    className="makeStyles-switchboxRoot-1"
  >
    <AugmentSwitch
      checked={false}
      iconUri="https://www.mousehuntgame.com/images/ui/hud/rift_valour/augments/highway_robbery.png"
      key="sigilHunter"
      label="Sigil Hunter"
      onChange={[Function]}
    />
    <AugmentSwitch
      checked={false}
      iconUri="https://www.mousehuntgame.com/images/ui/hud/rift_valour/augments/champion_pillage.png"
      key="secretResearch"
      label="Secret Research"
      onChange={[Function]}
    />
    <AugmentSwitch
      checked={false}
      iconUri="https://www.mousehuntgame.com/images/ui/hud/rift_valour/augments/secret_passage.png"
      key="superSiphon"
      label="Super Siphon"
      onChange={[Function]}
    />
    <AugmentSwitch
      checked={false}
      iconUri="https://www.mousehuntgame.com/images/ui/hud/rift_valour/augments/tower_ultimatum.png"
      key="ultimateUmbra"
      label="Ultimate Umbra"
      onChange={[Function]}
    />
    <AugmentSwitch
      checked={false}
      iconUri="https://www.mousehuntgame.com/images/ui/hud/rift_valour/augments/string_stepping.png"
      key="stringStepping"
      label="String Stepping"
      onChange={[Function]}
    />
  </div>
</Styled(MuiBox)>
`;
